# Procesamiento de Datos de Ventas

Enunciado:
Tienes un archivo de texto llamado ventas.txt que contiene información sobre las ventas de una tienda en el siguiente formato:

Tienes que realizar las siguientes tareas para procesar y analizar estos datos utilizando Python. Cada subtarea está asociada a un tema específico que se debe practicar:

## Subtarea 1: Leer y procesar el archivo
### Tema: Lectura de Archivos y Manejo de Datos

Leer el archivo y extraer los datos en formato adecuado (diccionario, lista, tupla, etc.). Cada línea del archivo representa una venta, que incluye producto, categoría, precio y cantidad. Guarda esta información en un diccionario donde la clave sea el nombre del producto y el valor sea una tupla con los detalles de cada venta.

In [7]:
def leer_datos(fichero):
    salida = {}
    with open(fichero,'r', encoding = 'utf-8') as file:
        next(file) ##  Nos saltamos la primera linea
        next(file) ##  Nos saltamos la segunda linea
        for linea in file: ## Iteramos linea por linea
            if linea.strip(): ## Por si nos encontramos una linea en blanco en medio
                producto, categoria, precio, unidades = linea.split(',') ## Guardamos en una variable cada zona unidad de la linea
                salida[producto] = (categoria,float(precio),int(unidades)) ## lo añadimos al diccionario como clave producto y el valor una tupla de lo restante
    return salida
data = leer_datos('ventas.txt')
data

{'Laptop': (' Electrónica', 800.0, 2),
 'Silla': (' Muebles', 150.0, 5),
 'Teclado': (' Electrónica', 50.0, 3),
 'Mesa': (' Muebles', 200.0, 1),
 'Auriculares': (' Electrónica', 120.0, 4),
 'Cámara': (' Fotografía', 400.0, 2),
 'Móvil': (' Electrónica', 600.0, 3),
 'Sofá': (' Muebles', 1000.0, 1),
 'Impresora': (' Electrónica', 250.0, 5),
 'Lámpara': (' Muebles', 80.0, 7),
 'Monitor': (' Electrónica', 300.0, 2),
 'Estante': (' Muebles', 120.0, 4),
 'Proyector': (' Electrónica', 450.0, 3),
 'Reloj': (' Joyería', 200.0, 6),
 'Pulsera': (' Joyería', 100.0, 8),
 'Cargador': (' Electrónica', 30.0, 10),
 'Altavoces': (' Electrónica', 150.0, 2),
 'Cámara de seguridad': (' Electrónica', 120.0, 4),
 'Bicicleta': (' Deportes', 350.0, 2),
 'Casco': (' Deportes', 50.0, 6),
 'Silla de oficina': (' Muebles', 200.0, 3),
 'Gafas de sol': (' Joyería', 120.0, 5),
 'Teclado inalámbrico': (' Electrónica', 60.0, 3),
 'Microondas': (' Electrodomésticos', 120.0, 2),
 'Tetera': (' Electrodomésticos', 35.0, 7)

## Ejercicio 1: Calcular el total de ventas por producto
### Nivel de dificultad: Básico

Escribe una función que reciba un diccionario con los productos, sus precios y cantidades, y calcule el total de ventas para cada producto. La salida debe ser un diccionario con los productos como claves y el total de ventas como valor.

Instrucciones:

- Usa un ciclo for para recorrer el diccionario.
- Para cada producto, multiplica el precio por la cantidad para obtener el total de ventas.
- Devuelve un diccionario con el producto y su total de ventas.

In [9]:
data = leer_datos('ventas.txt')
data
def ventas(data):
    return {k:v2*v3 for k,(v1,v2,v3) in data.items()} ## Hacemos un diccionario por compresion para sacar los elementos deseados
ventas(data)

{'Laptop': 1600.0,
 'Silla': 750.0,
 'Teclado': 150.0,
 'Mesa': 200.0,
 'Auriculares': 480.0,
 'Cámara': 800.0,
 'Móvil': 1800.0,
 'Sofá': 1000.0,
 'Impresora': 1250.0,
 'Lámpara': 560.0,
 'Monitor': 600.0,
 'Estante': 480.0,
 'Proyector': 1350.0,
 'Reloj': 1200.0,
 'Pulsera': 800.0,
 'Cargador': 300.0,
 'Altavoces': 300.0,
 'Cámara de seguridad': 480.0,
 'Bicicleta': 700.0,
 'Casco': 300.0,
 'Silla de oficina': 600.0,
 'Gafas de sol': 600.0,
 'Teclado inalámbrico': 180.0,
 'Microondas': 240.0,
 'Tetera': 245.0,
 'Aspiradora': 720.0,
 'Secador de pelo': 250.0,
 'Plancha': 210.0,
 'Nevera': 700.0,
 'Cafetera': 300.0}

## Ejercicio 2: Filtrar productos según su total de ventas
### Nivel de dificultad: Intermedio

Enunciado: Escribe una función que reciba un diccionario con productos y sus totales de ventas, y filtre aquellos productos cuyo total de ventas sea superior a un valor dado.

Instrucciones:

- Usa una función para calcular los totales de ventas, como en el ejercicio anterior.
- Filtra los productos que tienen un total de ventas superior a un valor dado usando un if dentro de un ciclo for.
- Devuelve un diccionario con los productos filtrados.

In [29]:
def total_ventas(data, valor):
    return { k:v for k,v in data.items() if valor < v}
data = leer_datos('ventas.txt')
data = ventas(data)
total_ventas(data, 500)

{'Laptop': 1600.0,
 'Silla': 750.0,
 'Cámara': 800.0,
 'Móvil': 1800.0,
 'Sofá': 1000.0,
 'Impresora': 1250.0,
 'Lámpara': 560.0,
 'Monitor': 600.0,
 'Proyector': 1350.0,
 'Reloj': 1200.0,
 'Pulsera': 800.0,
 'Bicicleta': 700.0,
 'Silla de oficina': 600.0,
 'Gafas de sol': 600.0,
 'Aspiradora': 720.0,
 'Nevera': 700.0}

## Ejercicio 3: Generar un informe de ventas con un resumen
### Nivel de dificultad: Avanzado

Enunciado: Escribe una función que reciba un diccionario de productos con sus precios y cantidades, y genere un informe con el total de ventas por producto, el total de ventas general y el producto más vendido. El informe debe ser una cadena de texto con el siguiente formato:


Informe de ventas:
Producto: Laptop | Total: 1600
Producto: Silla | Total: 750
...
Total general de ventas: 2830
Producto más vendido: Laptop (1600)

Instrucciones:

- Usa una función para calcular el total de ventas por producto.
- Calcula el total de ventas general sumando todos los totales de productos.
- Encuentra el producto más vendido mediante la función max().
- Devuelve el informe en formato de cadena.

In [41]:
def informe(productos):
    ventas_totales = 0
    mas_vendido = 0
    producto_mas_vendido = ''
    for producto,(v1,v2,v3) in productos.items():
        ventas_totales += v2*v3
        if v3 > mas_vendido: 
            mas_vendido = v3
            producto_mas_vendido = producto
    print(f'Total de ventas: {ventas_totales}')
    print(f'Producto mas vendido: {producto_mas_vendido}')

data = leer_datos('ventas.txt')
informe(data)

Total de ventas: 19145.0
Producto mas vendido: Cargador


## Subtarea 2: Filtrar productos de una categoría
### Tema: Condicionales (if, elif, else) y Bucles (for, while)

Filtrar los productos que pertenecen a la categoría "Electrónica" y calcula el total de ventas de estos productos (precio * cantidad).

## Subtarea 3: Encontrar el producto más caro y el más vendido
### Tema: Funciones y Funciones Lambda

Encontrar el producto más caro y el producto que más cantidad se vendió utilizando funciones y expresiones lambda. Usa las funciones max() para obtener los productos con los valores más altos en precio y cantidad, respectivamente.
python
Copiar


In [51]:
# {producto: (categoria, precio, unidades)}
def productos_destacados(datos):
    mas_caro = max(datos.items(), key = lambda x: x[1][1]) ## esta revisando el maximo valor del primer elemento del primer elemento [1][1]
    mas_vendido = max(datos.items(), key = lambda x: x[1][2]) ## producto mas vendido
    return mas_caro,mas_vendido
data = leer_datos('ventas.txt')
productos_destacados(data)

(('Sofá', (' Muebles', 1000.0, 1)), ('Cargador', (' Electrónica', 30.0, 10)))

## Subtarea 4: Filtrar productos usando listas por comprensión
### Tema: Listas por Comprensión

Filtrar productos cuyo precio sea mayor a 100 y hayan vendido más de 3 unidades. Utiliza listas por comprensión para obtener estos productos.

In [55]:
def productos_filtrados(datos):
    return [k for k,(v1,v2,v3) in datos.items() if v2 > 100 and v3 > 3]

data = leer_datos('ventas.txt')
productos_filtrados(data)

['Silla',
 'Auriculares',
 'Impresora',
 'Estante',
 'Reloj',
 'Cámara de seguridad',
 'Gafas de sol',
 'Aspiradora']

## Subtarea 5: Aplicar un descuento con map
### Tema: Funciones y map()

Aplicar un descuento del 10% a los productos cuya cantidad vendida sea mayor que 3. Utiliza la función map() para aplicar el descuento a todos los productos.

In [65]:
# {producto: (categoria,precio,unidades)}
def descuento(precio, cantidad):
    return (precio * 0.9 if cantidad > 3 else precio)
def aplicar_descuento(datos):
    return list(map(lambda x: (x[0], descuento(x[1][1], x[1][2])), datos.items()))
data = leer_datos('ventas.txt')
aplicar_descuento(data)

[('Laptop', 800.0),
 ('Silla', 135.0),
 ('Teclado', 50.0),
 ('Mesa', 200.0),
 ('Auriculares', 108.0),
 ('Cámara', 400.0),
 ('Móvil', 600.0),
 ('Sofá', 1000.0),
 ('Impresora', 225.0),
 ('Lámpara', 72.0),
 ('Monitor', 300.0),
 ('Estante', 108.0),
 ('Proyector', 450.0),
 ('Reloj', 180.0),
 ('Pulsera', 90.0),
 ('Cargador', 27.0),
 ('Altavoces', 150.0),
 ('Cámara de seguridad', 108.0),
 ('Bicicleta', 350.0),
 ('Casco', 45.0),
 ('Silla de oficina', 200.0),
 ('Gafas de sol', 108.0),
 ('Teclado inalámbrico', 60.0),
 ('Microondas', 120.0),
 ('Tetera', 31.5),
 ('Aspiradora', 162.0),
 ('Secador de pelo', 45.0),
 ('Plancha', 70.0),
 ('Nevera', 700.0),
 ('Cafetera', 100.0)]

## Subtarea 6: Filtrar productos con filter
### Tema: Función filter()

Filtrar productos cuyo nombre contenga la palabra "Teclado" utilizando filter().

In [69]:
def productos_teclado(datos): 
    return list(filter(lambda x: 'Teclado' in x[0], datos.items()))

data = leer_datos('ventas.txt')
productos_teclado(data)

[('Teclado', (' Electrónica', 50.0, 3)),
 ('Teclado inalámbrico', (' Electrónica', 60.0, 3))]

## Subtarea 7: Calcular el total de ventas con reduce
### Tema: Uso de reduce()

Calcular el total de ventas de todos los productos utilizando reduce(). La función reduce() permite acumular los resultados de manera eficiente, multiplicando el precio por la cantidad de cada producto.

In [83]:
# {producto: (categoria,precio,unidades)}
from functools import reduce
def total_ventas(datos):
    return reduce(lambda x,y: x+y[1][1] * y[1][2], datos.items(), 0)
total_ventas(leer_datos('ventas.txt'))

19145.0

## Subtarea 8: Medir el tiempo de ejecución con un decorador
### Tema: Decoradores

Medir el tiempo de ejecución de cada función utilizando un decorador. Crea un decorador que calcule y muestre el tiempo que tarda en ejecutarse cada función de procesamiento.

## Subtarea 9: Ordenar productos por ingresos usando yield
### Tema: Generadores y yield

Generar productos por ingresos, ordenando de mayor a menor por el total de ingresos (precio * cantidad). Utiliza un generador (yield) para devolver los productos en orden.

## Subtarea 10: Aplicar un descuento adicional con operadores ternarios
### Tema: Operadores Ternarios

Aplicar un descuento adicional del 5% a los productos cuyo total de ventas (precio * cantidad) supere los 2000. Utiliza operadores ternarios para determinar si el descuento debe aplicarse.

In [87]:
def aplicar_descuento(datos):
    return {producto:(v1,v2 * 0.95 if v2*v3 > 2000 else v2,v3) for producto, (v1,v2,v3) in datos.items()}
    #Diccionario por compresion con ternaria de por medio si se cumple la condicion
aplicar_descuento(leer_datos('ventas.txt'))

{'Laptop': (' Electrónica', 800.0, 2),
 'Silla': (' Muebles', 150.0, 5),
 'Teclado': (' Electrónica', 50.0, 3),
 'Mesa': (' Muebles', 200.0, 1),
 'Auriculares': (' Electrónica', 120.0, 4),
 'Cámara': (' Fotografía', 400.0, 2),
 'Móvil': (' Electrónica', 600.0, 3),
 'Sofá': (' Muebles', 1000.0, 1),
 'Impresora': (' Electrónica', 250.0, 5),
 'Lámpara': (' Muebles', 80.0, 7),
 'Monitor': (' Electrónica', 300.0, 2),
 'Estante': (' Muebles', 120.0, 4),
 'Proyector': (' Electrónica', 450.0, 3),
 'Reloj': (' Joyería', 200.0, 6),
 'Pulsera': (' Joyería', 100.0, 8),
 'Cargador': (' Electrónica', 30.0, 10),
 'Altavoces': (' Electrónica', 150.0, 2),
 'Cámara de seguridad': (' Electrónica', 120.0, 4),
 'Bicicleta': (' Deportes', 350.0, 2),
 'Casco': (' Deportes', 50.0, 6),
 'Silla de oficina': (' Muebles', 200.0, 3),
 'Gafas de sol': (' Joyería', 120.0, 5),
 'Teclado inalámbrico': (' Electrónica', 60.0, 3),
 'Microondas': (' Electrodomésticos', 120.0, 2),
 'Tetera': (' Electrodomésticos', 35.0, 7)